File: 15.3sum/main.py
# coding:utf-8
from typing import List


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        l = []
        nums.sort()
        nLen = len(nums)
        for i in range(nLen-1):
            if i != 0 and nums[i] == nums[i-1]:
                continue
            for (j, k) in self.twoSum(nums[i+1:], -nums[i]):
                l.append([nums[i], j, k])
        return l

    def twoSum(self, l, s):
        r = []
        i = 0
        j = len(l)-1
        while i < j:

            if l[i]+l[j] < s:
                i = i+1
            elif l[i]+l[j] > s:
                j = j-1
            else:
                a = l[i]
                b = l[j]
                r.append([a, b])
                while a == l[i] and i < j:
                    i = i+1
                while b == l[j] and i < j:
                    j = j-1
        return r


if __name__ == "__main__":
    print(Solution().threeSum([0, 0, 0, 0]))

File: 15.3sum/main.go
package main

import (
	"fmt"
	"sort"
)

func threeSum(nums []int) [][]int {
	l := [][]int{}
	sort.Ints(nums)
	fmt.Println(nums)
	n := len(nums)
	for i := 0; i < n-1; i++ {
		if i != 0 && nums[i] == nums[i-1] {
			continue
		}
		for _, v := range twoSum(nums[i+1:], -nums[i]) {
			l = append(l, []int{nums[i], v[0], v[1]})
		}
	}
	return l
}
func twoSum(nums []int, s int) [][]int {
	l := [][]int{}
	i := 0
	j := len(nums) - 1
	for i < j {
		if nums[i]+nums[j] < s {
			i++
		} else if nums[i]+nums[j] > s {
			j--
		} else {
			a := nums[i]
			b := nums[j]
			l = append(l, []int{a, b})
			for a == nums[i] && i < j {
				i++
			}
			for b == nums[j] && i < j {
				j--
			}
		}
	}
	return l
}
func main() {
	fmt.Println(threeSum([]int{0, 0, 0, 0, 0}))
}
